selfdrive/test/process_replay/process_replay.py

```diff
@@ -206,204 +191,291 @@ def get_car_params(msgs, fsm, can_sock, fingerprint):
-    fake_pubsubmaster=True,
+    polled_pubs=["liveLocationKalman"],
   ),
   ProcessConfig(
     proc_name="ubloxd",
-    pub_sub={
-      "ubloxRaw": ["ubloxGnss", "gpsLocationExternal"],
-    },
+    pubs=["ubloxRaw"],
+    subs=["ubloxGnss", "gpsLocationExternal"],
     ignore=["logMonoTime"],
+    config_callback=None,
     init_callback=None,
-    should_recv_callback=ublox_rcv_callback,
-    tolerance=None,
-    fake_pubsubmaster=False,
+    should_recv_callback=None,
   ),
   ProcessConfig(
     proc_name="laikad",
-    pub_sub={
-      "ubloxGnss": ["gnssMeasurements"],
-      "qcomGnss": ["gnssMeasurements"],
-    },
+    pubs=["ubloxGnss", "qcomGnss"],
+    subs=["gnssMeasurements"],
     ignore=["logMonoTime"],
-    init_callback=get_car_params,
+    config_callback=laikad_config_pubsub_callback,
+    init_callback=get_car_params_callback,
     should_recv_callback=None,
     tolerance=NUMPY_TOLERANCE,
-    fake_pubsubmaster=False,
     timeout=60*10,  # first messages are blocked on internet assistance
+    drained_pub="ubloxGnss", # config_callback will switch this to qcom if needed 
   ),
   ProcessConfig(
     proc_name="torqued",
-    pub_sub={
-      "liveLocationKalman": ["liveTorqueParameters"],
-      "carState": [], "controlsState": [],
-    },
+    pubs=["liveLocationKalman", "carState", "carControl"],
+    subs=["liveTorqueParameters"],
     ignore=["logMonoTime"],
-    init_callback=get_car_params,
+    config_callback=None,
+    init_callback=get_car_params_callback,
     should_recv_callback=torqued_rcv_callback,
     tolerance=NUMPY_TOLERANCE,
-    fake_pubsubmaster=True,
+    polled_pubs=["liveLocationKalman"],
   ),
 ]
 
 
 def replay_process(cfg, lr, fingerprint=None):
   with OpenpilotPrefix():
-    if cfg.fake_pubsubmaster:
-      return python_replay_process(cfg, lr, fingerprint)
+    controlsState = None
+    initialized = False
+    for msg in lr:
+      if msg.which() == 'controlsState':
+        controlsState = msg.controlsState
+        if initialized:
+          break
+      elif msg.which() == 'carEvents':
+        initialized = car.CarEvent.EventName.controlsInitializing not in [e.name for e in msg.carEvents]
+
+    assert controlsState is not None and initialized, "controlsState never initialized"
+
+    CP = [m for m in lr if m.which() == 'carParams'][0].carParams
+    if fingerprint is not None:
+      setup_env(cfg=cfg, controlsState=controlsState, lr=lr, fingerprint=fingerprint)
     else:
-      return replay_process_with_sockets(cfg, lr, fingerprint)
+      setup_env(CP=CP, cfg=cfg, controlsState=controlsState, lr=lr)
+
+    if cfg.config_callback is not None:
+      params = Params()
+      cfg.pubs, cfg.drained_pub = cfg.config_callback(params, cfg)
+      cfg.polled_pubs = list(set(cfg.polled_pubs) & set(cfg.pubs))
+
+    all_msgs = sorted(lr, key=lambda msg: msg.logMonoTime)
+    pub_msgs = [msg for msg in all_msgs if msg.which() in set(cfg.pubs)]
+
+    with ReplayContext(cfg) as rc:
```

Comment:
the context managers can be chained, so the top can just be `with OpenpilotPrefix(), with ReplayContext(cfg) as rc:`

Reply:
Done

Reply:
Ok, it must be the way it was, because the config_callback must be run before ReplayContext is created, because it changes the `pubs` and `drained_pub` of the process - laikad uses that to switch between qcom and ublox